objcgen/objcgenerator.cs

```diff
@@ -102,36 +103,75 @@ protected override void Generate (Type t)
 			implementation.WriteLine ($"\tif (!{managed_name}_class) {{");
 			implementation.WriteLine ("\t\t__initialize_mono ();");
 			implementation.WriteLine ("\t\t__lookup_assembly_managed ();");
-			implementation.WriteLine ($"\t\t{managed_name}_class = mono_class_from_name (__{t.Assembly.GetName ().Name}_image, \"\", \"{managed_name}\");");
+			implementation.WriteLine ($"\t\t{managed_name}_class = mono_class_from_name (__{t.Assembly.GetName ().Name}_image, \"{t.Namespace}\", \"{managed_name}\");");
 			implementation.WriteLine ("\t}");
 			implementation.WriteLine ("}");
+			implementation.WriteLine ();
 
 			var native_name = GetTypeName (t);
 			headers.WriteLine ($"// {t.AssemblyQualifiedName}");
 			headers.WriteLine ($"@interface {native_name} : {GetTypeName (t.BaseType)} {{");
 			headers.WriteLine ("\tMonoEmbedObject* _object;");
 			headers.WriteLine ("}");
-
 			headers.WriteLine ();
-			if (t.IsSealed && t.IsAbstract) {
-				// don't allow instantiation of static types from ObjC code
-				headers.WriteLine ("// a .net static type cannot be initialized");
-				headers.WriteLine ("- (instancetype)init NS_UNAVAILABLE;");
-				headers.WriteLine ();
-			}
 
 			implementation.WriteLine ($"// {t.AssemblyQualifiedName}");
-			implementation.WriteLine ($"@implementation {managed_name}");
+			implementation.WriteLine ($"@implementation {native_name}");
 			implementation.WriteLine ();
 
+			var default_init = false;
+			foreach (var ctor in t.GetConstructors ()) {
+				// .cctor not to be called directly by native code
+				if (ctor.IsStatic)
+					continue;
+				if (!ctor.IsPublic)
+					continue;
+				if (ctor.ParameterCount == 0) {
+					headers.WriteLine ("- (instancetype)init;");
+
+					implementation.WriteLine ("- (instancetype)init");
+					implementation.WriteLine ("{");
+					implementation.WriteLine ($"\tconst char __method_name [] = \"{t.FullName}:.ctor()\";");
+					implementation.WriteLine ("\tstatic MonoMethod* __method = nil;");
+					implementation.WriteLine ("\tif (!__method) {");
+					implementation.WriteLine ($"\t\t__lookup_class_{managed_name} ();");
+					implementation.WriteLine ($"\t\t__method = mono_embeddinator_lookup_method (__method_name, {managed_name}_class);");
+					implementation.WriteLine ("\t}");
+					implementation.WriteLine ("\tif (self = [super init]) {");
```

Comment:
This won't work for managed classes that inherit from NSObject (in which case we should just call the managed constructor instead of `[super init]`).